### DISCUSSION TITLE
Is it possible to do manual rebuilds with changed files (Node API)?

### DISCUSSION THREAD

**Discussion by Exigerr** [webpack/webpack#17982] [Q&A] [answered]
Hi,

I've got a slightly complicated setup with multiple projects and packages which means that I need to be in full control of the order of builds and rebuilds. That means that I cannot use the Webpack file watcher (`compiler.watch`).

The `compiler`'s `run` method can be called multiple times though. 
If you are using the cache (memory) option, then trigger a second build sometime much later on, the build will run however no changes to files will be picked up (uses the cache as expected).

In my situation, fortunately I'm fully aware of what files have changed.
Is there a simple way to run a build, then invalidate specific cached files, then trigger another build?

Essentially what I'm looking for is to be able to:
- Call `compiler.run`
- Wait for build to finish
- *Sometime later a file changes*
- Do my processing
- Tell the `compiler` reference to invalidate its cache for the changed file(s)
- Call `compiler.run` again (uses cache for unchanged files)

Disabling the cache is an option, but would lose out quick build times.
I have noticed that the `compiler` has a `cache` property and `getCache` method, however there seems to be very little information / documentation about how to use them or if invalidating specific files is possible.

**TLDR:** Need to be able to do incremental rebuilds but without using the Webpack file watcher.

**Comment by alexander-akait** [comment]
> Tell the compiler reference to invalidate its cache for the changed file(s)

It can be done, but it's not easy, why do you need? I feel you are doing something wrong

**Reply by Exigerr** [reply]
We have our own build systems for individual packages, performant file watcher which spans across projects and apps, and coordinating system to tell when each thing should do what. 

One example flow could be:
- File changes in a package
- Another file changes in a different package
- Our existing file watcher triggers (watching things cross packages / apps / projects)
- Our system sees that there is a dependency relationship between the packages, so it builds them in the correct order
   - For some older packages, post-processing can occur on the `dist` file content
- Our system sees everything has settled, knows that the packages are related to X applications, so it goes to safely trigger a rebuild on those applications.
 
Trying to fit Webpack 5 in for application builds without having to fully redo our existing processes.

Originally we were using Rollup which has a cache and rebuild mechanism. After a build you get a simple cache which you can quickly find a file/module by absolute path and then remove it. Triggering a build after that allows it to reuse the cache for everything else, but read from disk and re-process anything missing from cache.

Unfortunately Rollup is rather slow on very large projects (bundle time + sourcemaps). From our tests Webpack 5 (in development and production mode) seems to be a little faster and much much faster when it rebuilds via its watcher.

Webpack 5's watcher however doesn't provide enough granular control. From the looks of it you can't simply tell it, 'ignore this file change because something else is working, okay that is done now you can run'.

With the `watchOptions` you can provide a delay or ignore things entirely. we need more of a deferred execution.
If we were to just use Webpack's `watch` we would either have some unnecessary builds trying to run while things elsewhere are changing, or a consistent slow delay before it kick-starts.

Additionally it would not be super optimal to run multiple file watchers (running Webpack's on top of the one we are using already which covers what we need to care about).

**Comment by Exigerr** [comment]
Not had much luck with the `cache` / `getCache` API.

I have found a potential way to deal with this but unsure if there is a better solution.
- Create a custom plugin which makes use of `stillValidModule` hook (`compilation.hooks.stillValidModule`).
- Update the plugin reference with a list of files which it should care about
- When the next build is manually triggered (with memory cache), `stillValidModule` will be called for the cached files.
- If the module provided to the `stillValidModule` hook matches one of the files, then call `compilation.rebuildModule` with it.

From quick tests that seems to correctly get Webpack to grab the new file and rebuild the chunks as needed.

Not sure if there is a better hook or not.
`stillValidModule` is listed in the types and noted on release notes, however there seems to be little to no information about it or what it should be used for.
Source code only shows that it is called after `needBuild` and it says it doesn't need to build the module again.

**Reply by alexander-akait** [reply, marked as answer]
We have https://github.com/webpack/webpack/blob/main/tooling/print-cache-file.js to debug purposes, but I don't think it really helps to solve the problem.

But I see what you want to achieve...

We have https://github.com/webpack/webpack/blob/main/lib/NormalModule.js#L265, so you can implement an own logic when you need rebuild, did you try it?

**Reply by Exigerr** [reply]
Apologies for the slow reply.
I had tried the `NormaModule` `needBuild` hook previously and noticed it wasn't being called. After your comment, I went back and investigated further....silly me found that we still had the `speed-measure-webpack-v5-plugin` in use. That unfortunately breaks reference equality on the `compilation` object (hooks would never be called on the right thing due how `getCompilationHooks` works).

After removing the `speed-measure-webpack-v5-plugin` then retesting the `needBuild` hook again, it worked perfectly and I can easily make use of it for our use case!
Thank you very much for the assistance.
